### PR TITLE
feat: [M12] Direct-tsserver semanticDiagnosticsSync path to bypass the LSP client debounce

### LINKED ISSUE #279: [M12] Direct-tsserver semanticDiagnosticsSync path to bypass the LSP client debounce
Part of #198 (M12). **Unblocked** — #278 is closed and its diagnosis is established.

## Capability

*A person can get TypeScript diagnostics for a warm, multi-file project at a latency set by
tsserver's real recheck cost (~10–30 ms), not by `typescript-language-server`'s hardcoded
client-side debounce (~350 ms).*

## Why

#278 established that ~350 ms of the 378 ms `diagnostics` latency measured in #220 / PR #277 is
fixed client-side debounce inside `typescript-language-server` 5.3.0 (`cli.mjs:17868`'s
`min(max(ceil(lineCount/20), 300), 800)` plus `:20516`'s 50 ms `pDebounce`), and that the genuine
recheck cost is **~11–31 ms** — inside the 50 ms bar the type-aware eval arm needs.

Neither timer is configurable, and `typescript-language-server` exposes no synchronous path:
`semanticDiagnosticsSync`, `syntacticDiagnosticsSync`, `suggestionDiagnosticsSync` and
`geterrForProject` appear in the bundle **only as enum constants** in the vendored tsserver protocol
types (`cli.mjs:14806`) — the LSP layer never invokes them.

So realising that ~30 ms means bypassing `typescript-language-server` and speaking **tsserver's own
protocol** directly: spawn `node_modules/typescript/lib/tsserver.js` (the pinned `typescript`
5.9.3), `open` / `updateOpen`, then `semanticDiagnosticsSync`, which replies synchronously with no
debounce. This is a second transport, so it earns its own module rather than a flag on
`TsLspService`.

## In scope

- `src/lsp/ts_server_direct.py` — a new above-the-seam, stdlib-only client speaking tsserver's own
  protocol (newline-delimited JSON in, `Content-Length`-framed JSON out; `request_seq` correlation,
  not LSP `id`).
- Its diagnostics op: `syntacticDiagnosticsSync` + `semanticDiagnosticsSync`, merged, mapped to the
  existing `Diagnostic` dataclass with the same two traps `ts_lsp.py` documents (`TS{code}` string
  codes; errors only).
- A parity gate against `TsLspService` over the #194 error-injection set.
- A third `diagnostics_direct` column in `scripts/bench_ts_lsp.py`, and a regenerated
  `results/ts_lsp_bench.json`.
- Recording the measured verdict in `docs/design/12-lsp-in-the-loop.md`, including the case where
  the answer is "this did not help".

## Out of scope

- Any change to `TsLspService`'s behavior. It stays as the LSP-conformant client; `completions` (the
  op #226 actually needs) already runs at ~2,500 calls/s, so nothing about #226 is affected.
- The other five ops (`definition`/`references`/`completions`/`quickinfo`/`documentSymbols`) over the
  direct transport. Diagnostics is the op the debounce hurts.
- Any patching of the vendored `cli.mjs`.
- A Swift port of this transport (a follow-up if the number justifies it; see design/12's
  "Boundary with #279").
- Switching any existing consumer (`oracle.py`, `harness.py`, the eval arms) onto the new client.
  This issue delivers the client, its gate, and the measurement; adoption is a separate decision.

## Acceptance criteria

1. `src/lsp/ts_server_direct.py` exists, is stdlib-only, and is listed in
   `tests/test_import_guard.py`'s `PORTABLE_MODULES`.
2. It spawns the pinned `node_modules/typescript/lib/tsserver.js` via a `resolve_tsserver()` that
   mirrors `tsc.resolve_tsc()` (returns `None`, never raises, on a host with no toolchain).
3. It exposes `open_project(files)` / `update(path, text)` / `diagnostics(path) -> List[Diagnostic]`,
   matching `TsLspService`'s signatures for those three so the two are directly comparable.
4. `diagnostics()` never raises — on timeout, a dead child, or a `success: false` reply it counts the
   event and returns `[]`, matching `TsLspService`'s documented contract.
5. **Parity**: over every record in `eval_sets/ts_error_injection/eval.jsonl`, the direct client's
   diagnostics equal `TsLspService`'s, compared as the sorted set of `(code, line, col)`. Any
   systematic divergence is either fixed or documented in design/12 as a known, characterized
   difference — a faster oracle that silently disagrees is not a win.
6. **Measurement**: `scripts/bench_ts_lsp.py` reports a `diagnostics_direct` op alongside
   `diagnostics`, on the same 5,000-file project and the same alternating introduce/revert edit
   cycle, and `results/ts_lsp_bench.json` is regenerated with it.
7. The measured `diagnostics_direct` median is recorded against the 50 ms bar and a
   `verdict_direct` field is emitted; `docs/design/12-lsp-in-the-loop.md` states plainly whether the
   type-aware eval arm's verdict changes from WARN, including if it does not.
8. Binary-free mechanism tests (`tests/test_ts_server_direct_mechanism.py`, the `os.pipe()` pattern
   from `tests/test_jsonrpc.py`) are the CI gate; live toolchain-gated tests carry parity.
9. Every statement this change falsifies is corrected in place — the `#279 (open, not built here)`
   and `would sidestep` framings in `docs/design/12-lsp-in-the-loop.md` and
   `src/lsp/ts_service.py`'s module docstring.
10. `docs/feature-matrix.md` and `docs/test-plan.md` carry a row for this capability, updated in the
    same commits as the code.

## Edge cases

- **Framing asymmetry**: tsserver reads newline-delimited JSON on stdin but writes
  `Content-Length`-framed JSON on stdout. Reusing `jsonrpc.encode_message` for writes would hang.
- **Event vs response**: tsserver interleaves unsolicited `type: "event"` messages
  (`projectLoadingStart/Finish`, `telemetry`, `configFileDiag`, `syntaxDiag`/`semanticDiag`) with
  `type: "response"` messages correlated by `request_seq`. Events must never resolve a waiter.
- **`success: false`** responses and responses with no `body`.
- **Cold project load**: the first `open` blocks on `projectLoadingFinish`; that cost is cold-start,
  not per-recheck, and must not contaminate the timed loop.
- **Semantic vs syntactic split**: `semanticDiagnosticsSync` alone misses parse errors the LSP layer
  reports. Both must be merged for parity.
- **`category` filtering**: tsserver reports `"error"` / `"warning"` / `"suggestion"`; only
  `"error"` survives, matching `ts_lsp.py`'s Trap B.
- **Code formatting**: tsserver's `code` is a bare integer; it must become `f"TS{code}"` or
  `is_incomplete` breaks (`ts_lsp.py`'s Trap A).
- **Coordinates**: tsserver's protocol is 1-based `{line, offset}`; LSP is 0-based
  `{line, character}`. Both must land on `Diagnostic`'s 1-indexed convention.
- **Full-text update**: `updateOpen` takes `textChanges`, not whole files; a whole-file replace must
  be expressed correctly (or via re-`open`) and proved by the parity gate.
- **Dead child / EOF** mid-request must fail the waiter, not hang.
- **No toolchain**: every live test and the bench column skip cleanly, as `resolve_ts_lsp() is None`
  already does elsewhere.

## Verification

- `.venv/bin/python -m pytest tests/test_ts_server_direct_mechanism.py tests/test_ts_server_direct.py -q`
- `.venv/bin/python -m pytest tests/test_import_guard.py -q`
- `.venv/bin/python -m pytest -q -rs` (full suite; nothing unrelated broke)
- `.venv/bin/python scripts/bench_ts_lsp.py --n-files 5000 --out results/ts_lsp_bench.json`
- Read `docs/design/12-lsp-in-the-loop.md` and confirm no statement still says #279 is unbuilt.

## Tasks

- [x] `src/lsp/ts_server_direct.py` — tsserver-protocol client (seq/command framing, not LSP JSON-RPC)
- [x] `semanticDiagnosticsSync` + `syntacticDiagnosticsSync` diagnostics path
- [x] **Parity gate**: identical diagnostics to `TsLspService` over the #194 error-injection set (`eval_sets/ts_error_injection/eval.jsonl`)
- [x] Add it as a third column in `scripts/bench_ts_lsp.py`; confirm the ~30 ms projection empirically
- [x] Decide from the measured number whether the type-aware eval arm's verdict changes from WARN
- [x] Add to `PORTABLE_MODULES` in `tests/test_import_guard.py` (stdlib only, above the seam)
- [x] Binary-free mechanism tests (the `tests/test_jsonrpc.py` `os.pipe()` pattern) — the real CI gate
- [x] Correct design/12 + `ts_service.py`'s docstring in place; add the feature-matrix and test-plan rows

## The decision this issue must not dodge

`TsLspService` stays regardless: it is the LSP-conformant client, and `completions` already runs
fast enough that **nothing about #226 is blocked**. This is purely about the type-aware eval arm's
per-candidate edit→recheck cycle. If that arm ends up batching candidates rather than checking them
one at a time, the 350 ms amortises and this work buys little. Criterion 7 exists so that outcome is
recorded honestly rather than reframed.

### PR DESCRIPTION
**Pull request by travisgalloway**
Closes #279

## Summary

A **second transport** to the same pinned TypeScript, speaking **tsserver's own protocol** to `node_modules/typescript/lib/tsserver.js` instead of LSP to `typescript-language-server`, so #278's two hardcoded client-side timers are not on the diagnostics path at all.

- **`src/lsp/ts_server_direct.py`** (new, above the seam, stdlib only) — `resolve_tsserver()`, `TsServerEndpoint` (framing + `request_seq` demux), `TsServerDirect` (`open_project`/`update`/`diagnostics`, signature-matched to `TsLspService`). It is a separate module rather than a flag on `jsonrpc.py` because the wire protocol is genuinely asymmetric: tsserver reads **newline-delimited JSON** on stdin but writes **`Content-Length`-framed** JSON on stdout — so `jsonrpc.read_message` is *reused unchanged* and `encode_message` is *not* (a header on stdin hangs the child silently). Responses correlate on `request_seq`, not `id`; unsolicited events interleave; the payload is 1-based `{line, offset}` with a string `category`.
- **Second, quieter win:** the `*Sync` commands always answer, so `[]` here is an *answer* rather than a possible `n_no_publish` (#278's `cli.mjs:20524` ambiguity). There is deliberately no such counter on this class.
- **Bench third column** — `scripts/bench_ts_lsp.py` gains `diagnostics_direct` on the *same* generated project, the *same* introduce/revert rotation, the *same* iteration count, with its own BLIND guard (`_sanity_probe_direct`). `verdict` keeps its meaning (the LSP-conformant client, the #277/#278 record); `verdict_direct` is a separate counterfactual through the same `_verdict` function.

## The measured number

`--n-files 5000 --warmup 20 --iters 200`, seed 0:

| op | median | p95 | calls/s | non-empty |
|---|---:|---:|---:|---|
| `diagnostics` (LSP client) | 377.0 ms | 395.4 ms | 3.1 | 32% — `introduce` 64% |
| `diagnostics_direct` (#279) | **19.7 ms** | 22.6 ms | 50.2 | 50% — `introduce` **100%** |

`verdict` **WARN** (unchanged), `verdict_direct` **PASS**. 19.7 ms clears the type-aware arm's 50 ms bar with a 2.5× margin, confirming #278's ~11–33 ms projection from the other direction.

Stated plainly in design/12, both ways: latency is no longer the blocker — **but** nothing consumes `TsServerDirect` yet and the arm may batch candidates for unrelated reasons, so this does not by itself unblock anything. Adoption is a separate decision.

Also recorded honestly: the direct column observes **100%** of introduced `TS2339`s where the LSP client observes **64%** (#278's residual in-flight notification race). That is a correctness difference in the large-warm-program regime, not only a latency one — and the 96-record parity gate does *not* see it, because single-file candidates coalesce tsserver's two passes into one push.

## Testing

- [x] Tests added/updated — `tests/test_ts_server_direct_mechanism.py` (33 tests, binary-free `os.pipe()` scripted server; **passes with no `node` on PATH** — verified with `env PATH=/usr/bin:/bin`) covering the framing asymmetry, `request_seq` correlation, events never resolving waiters, timeout/EOF/dead-child/refused-command handling, and every mapping trap. `tests/test_ts_server_direct.py` (12 tests, toolchain-gated) carries the **parity gate**: sorted `(code, line, col)` sets identical to `TsLspService` over all **96** #194 records in **both** directions (192 candidates, **0 mismatches**), plus the anti-vacuity half asserting the direct client independently reports each record's `expected_diagnostic` — parity against a silent oracle would be vacuous.
- [x] All tests passing — `.venv/bin/python -m pytest -q -rs`: **1640 passed, 12 skipped** (6:13). Import guard green with `src.lsp.ts_server_direct` added to `PORTABLE_MODULES`.
- [x] Manual testing completed — full bench re-run and `results/ts_lsp_bench.json` regenerated and committed.

## Docs (same commit as the code)

- `docs/design/12-lsp-in-the-loop.md` — new "The direct-`tsserver` transport (#279)" subsection with the measurement and the criterion-7 decision, plus **six in-place corrections** of statements this change falsified (no "superseded" annotations).
- `src/lsp/ts_service.py` module docstring and `src/lsp/__init__.py` now name the landed module.
- `docs/feature-matrix.md` **LSP-10** and `docs/test-plan.md` **LSP-10**, keyed identically.
- `docs/parked-findings.md` — 3 out-of-scope discoveries logged, none acted on.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_011ur4BdaFTqnwqSUAx5TrX8

### CODE REVIEW

File: src/lsp/ts_server_direct.py [hunk 1]
```diff
@@ -0,0 +1,616 @@
...
+        self._abs_path(path).write_text(text, encoding="utf-8")
+        self._files[path] = text
+        self._versions[path] += 1
+        self._request_retrying("updateOpen", {"openFiles": [{
+            "file": self._file(path),
+            "fileContent": text,
+            "scriptKindName": "TS",
+            "projectRootPath": str(self.scratch_dir),
+        }]})
+        return self._versions[path]
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply src/lsp/ts_server_direct.py:545]: VALID — fixed in 45c581d. `update()` now catches the `updateOpen` failure and counts it into the same `n_timeouts` / `n_command_errors` buckets `diagnostics()` uses, so it keeps `TsLspService.update()`'s non-raising contract. Both branches are pinned by new tests in `tests/test_ts_server_direct_mechanism.py`.

File: src/lsp/ts_server_direct.py [hunk 2]
```diff
@@ -0,0 +1,616 @@
...
+import tempfile
+import threading
+import time
+from collections import deque
+from pathlib import Path
+from typing import Any, BinaryIO, Callable, Deque, Dict, List, Mapping, Optional, Set, Tuple
+
+from .diagnostics import Diagnostic, line_col_to_offset
+from .jsonrpc import _Waiter, read_message
+from .tsc import DEFAULT_TSCONFIG_PATH, SET_DIR
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply src/lsp/ts_server_direct.py:76]: VALID — fixed in 45c581d. `_Waiter` is now the public `jsonrpc.Waiter`: `ts_server_direct.py` reuses the rendezvous primitive verbatim for tsserver's `request_seq`-keyed demux, so it is a real cross-module type rather than a private detail of `jsonrpc.py`. Docstring says so.

File: scripts/bench_ts_lsp.py [hunk 3]
```diff
@@ -488,10 +560,43 @@ def main() -> int:
...
+                # BLIND applies to the third column too -- an unchecked fast
+                # zero is worse than no column at all.
+                print("BLIND: direct-tsserver sanity probe failed -- nothing was measured.")
+                for name, ok in direct_probe["checks"].items():
+                    print(f"  {name}: {'OK' if ok else 'FAILED'}")
+                _write_results(out_path, {"summary": {"verdict": "BLIND",
+                                                      "sanity_probe": probe,
+                                                      "direct_sanity_probe": direct_probe},
+                                          "ops": {}})
+                raise SystemExit(2)
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply scripts/bench_ts_lsp.py]: VALID — fixed in 45c581d. Both BLIND early exits now go through `_full_summary()` and emit the complete `summary` schema. A truncated dict on the failure path — the one most likely to be inspected by tooling afterwards — would silently break any consumer assuming a stable schema.
